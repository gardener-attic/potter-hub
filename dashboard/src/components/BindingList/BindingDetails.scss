.BindingDetails code {
  word-wrap: break-word;
}
