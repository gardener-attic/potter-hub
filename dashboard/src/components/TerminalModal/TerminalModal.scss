.MessageDetails {
  display: flex;
  align-items: center;
  justify-content: center;
  white-space: pre-wrap;
}
