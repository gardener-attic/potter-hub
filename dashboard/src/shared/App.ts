import { axiosWithAuth } from "./AxiosInstance";
import { hapi } from "./hapi/release";
import { HELM_URL } from "./Kube";
import { IAppOverview, IChartVersion } from "./types";

export class App {
  public static getResourceURL(namespace?: string, name?: string, query?: string) {
    let url = HELM_URL;
    if (namespace) {
      url += `/namespaces/${namespace}`;
    }
    url += "/releases";
    if (name) {
      url += `/${name}`;
    }
    if (query) {
      url += `?${query}`;
    }
    return url;
  }

  public static async create(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = App.getResourceURL(namespace);
    const { data } = await axiosWithAuth.post(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async rollback(releaseName: string, namespace: string, revision: number) {
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(releaseName: string, namespace: string, keepHistory: boolean) {
    let keepHistoryQuery;
    if (keepHistory) {
      keepHistoryQuery = "keepHistory=true";
    }
    const { data } = await axiosWithAuth.delete(
      App.getResourceURL(namespace, releaseName, keepHistoryQuery),
    );
    return data;
  }

  public static async listApps(namespace?: string, allStatuses?: boolean) {
    let query;
    if (allStatuses) {
      query = "statuses=all";
    }
    const { data } = await axiosWithAuth.get<{ data: IAppOverview[] }>(
      App.getResourceURL(namespace, undefined, query),
    );
    return data.data;
  }

  public static async getRelease(namespace: string, name: string) {
    const { data } = await axiosWithAuth.get<{ data: hapi.release.Release }>(
      this.getResourceURL(namespace, name),
    );
    return data.data;
  }
}
