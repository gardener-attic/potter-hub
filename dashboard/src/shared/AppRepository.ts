import { axiosWithAuth } from "./AxiosInstance";
import { UI_BACKEND_ROOT_URL } from "./Kube";
import { IAppRepository, IAppRepositoryList } from "./types";

export class AppRepository {
  public static async list(namespace: string) {
    const { data } = await axiosWithAuth.get<IAppRepositoryList>(
      AppRepository.getResourceLink(namespace),
    );
    return data;
  }

  public static async get(name: string, namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async update(name: string, namespace: string, newApp: IAppRepository) {
    const { data } = await axiosWithAuth.put(AppRepository.getSelfLink(name, namespace), newApp);
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const { data } = await axiosWithAuth.delete(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async create(
    name: string,
    namespace: string,
    url: string,
    auth: any,
    syncJobPodTemplate: any,
  ) {
    const { data } = await axiosWithAuth.post<IAppRepository>(
      AppRepository.getResourceLink(namespace),
      {
        apiVersion: "kubeapps.com/v1alpha1",
        kind: "AppRepository",
        metadata: {
          name,
        },
        spec: { auth, type: "helm", url, syncJobPodTemplate },
      },
    );
    return data;
  }

  private static APIBase: string = UI_BACKEND_ROOT_URL;
  private static APIEndpoint: string = `${AppRepository.APIBase}/apprepositories`;
  
  private static getResourceLink(namespace?: string): string {
    return AppRepository.APIEndpoint;
  }

  private static getSelfLink(name: string, namespace: string): string {
    return `${AppRepository.APIEndpoint}/${name}`;
  }
}
