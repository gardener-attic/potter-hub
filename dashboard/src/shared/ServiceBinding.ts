import { axiosWithAuth } from "./AxiosInstance";
import { K8S_REVERSE_PROXY_URL } from "./Kube";
import { definedNamespaces } from "./Namespace";
import { ICondition, ServiceCatalog } from "./ServiceCatalog";

interface IK8sApiSecretResponse {
  kind: string;
  apiVersion: string;
  metadata: {
    selfLink: string;
    resourceVersion: string;
  };
  data: { [s: string]: string };
}

interface IServiceBinding {
  metadata: {
    name: string;
    selfLink: string;
    uid: string;
    resourceVersion: string;
    creationTimestamp: string;
    finalizers: string[];
    generation: number;
    namespace: string;
  };
  spec: {
    externalID: string;
    instanceRef: {
      name: string;
    };
    secretName: string;
  };
  status: {
    conditions: ICondition[];
    asyncOpInProgress: boolean;
    currentOperation: string;
    reconciledGeneration: number;
    operationStartTime: string;
    externalProperties: {};
    orphanMitigationInProgress: boolean;
    unbindStatus: string;
  };
}

export interface IServiceBindingWithSecret {
  binding: IServiceBinding;
  secret?: IK8sApiSecretResponse;
}

export class ServiceBinding {
  public static async create(
    bindingName: string,
    instanceRefName: string,
    namespace: string,
    parameters: {},
  ) {
    const url = ServiceBinding.getLink(namespace);
    const { data } = await axiosWithAuth.post<IServiceBinding>(url, {
      metadata: {
        name: bindingName,
      },
      spec: {
        instanceRef: {
          name: instanceRefName,
        },
        parameters,
      },
    });
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const url = this.getLink(namespace, name);
    return axiosWithAuth.delete(url);
  }

  public static async get(namespace: string, name: string) {
    const url = this.getLink(namespace, name);
    const { data } = await axiosWithAuth.get<IServiceBinding>(url);
    return data;
  }

  public static async list(namespace?: string): Promise<IServiceBindingWithSecret[]> {
    const bindings = await ServiceCatalog.getItems<IServiceBinding>("servicebindings", namespace);

    return Promise.all(
      bindings.map(binding => {
        const { secretName } = binding.spec;
        const ns = binding.metadata.namespace;
        return axiosWithAuth
          .get<IK8sApiSecretResponse>(this.secretEndpoint(ns) + secretName)
          .then(response => {
            return { binding, secret: response.data };
          })
          .catch(err => {
            // return with undefined secrets
            return { binding };
          });
      }),
    );
  }

  private static getLink(namespace?: string, name?: string): string {
    return `${K8S_REVERSE_PROXY_URL}/apis/servicecatalog.k8s.io/v1beta1${
      namespace ? `/namespaces/${namespace}` : ""
    }/servicebindings${name ? `/${name}` : ""}`;
  }

  private static secretEndpoint(namespace: string = definedNamespaces.default): string {
    return `${K8S_REVERSE_PROXY_URL}/api/v1/namespaces/${namespace}/secrets/`;
  }
}
