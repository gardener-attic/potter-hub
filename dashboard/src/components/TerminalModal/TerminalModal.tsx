import * as React from "react";
import Modal from "react-modal";

import "./TerminalModal.css";

interface ITerminalModalProps {
  modalIsOpen: boolean;
  closeModal: () => Promise<any>;
  title: string;
  message: string;
}

const TerminalModal: React.SFC<ITerminalModalProps> = props => {
  return (
    <div className="MessageDetails">
      <Modal
        isOpen={props.modalIsOpen}
        onRequestClose={props.closeModal}
        contentLabel="Modal"
        className="Terminal"
        style={{
          overlay: {
            display: "flex",
            alignItems: "center",
            justifyContent: "center",
          },
          content: {
            maxWidth: "80%",
          },
        }}
      >
        <div className="Terminal__Top type-small">
          <div className="Terminal__Top__Buttons">
            <a>
              <span
                className="Terminal__Top__Button Terminal__Top__Button--red"
                onClick={props.closeModal}
              />
            </a>
          </div>
          <div className="Terminal__Top__Title">{props.title}</div>
        </div>
        <div className="Terminal__Tab">
          <pre className="Terminal__Code">
            <code>{props.message}</code>
          </pre>
        </div>
      </Modal>
    </div>
  );
};

export default TerminalModal;
