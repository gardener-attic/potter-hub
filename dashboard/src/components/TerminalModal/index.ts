import TerminalModal from "./TerminalModal";

export default TerminalModal;
