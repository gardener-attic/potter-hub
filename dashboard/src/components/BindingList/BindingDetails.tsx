import * as React from "react";

import TerminalModal from "../../components/TerminalModal";
import { IServiceBindingWithSecret } from "../../shared/ServiceBinding";
import "./BindingDetails.css";

interface IBindingDetailsState {
  modalIsOpen: boolean;
}

class BindingDetails extends React.Component<IServiceBindingWithSecret, IBindingDetailsState> {
  public state = {
    modalIsOpen: false,
  };

  public render() {
    const { binding, secret } = this.props;
    const { secretName } = binding.spec;

    let secretDataArray: string[][] = [];
    if (secret) {
      const secretData = Object.keys(secret.data).map(k => [k, atob(secret.data[k])]);
      secretDataArray = [...secretData];
    }
    let message = "";
    if (secretDataArray.length > 0) {
      message = "";
      secretDataArray.forEach(statusPair => {
        const [key, value] = statusPair;
        message += `${key}: ${value}\n`;
      });
    } else {
      message = "The secret is empty";
    }
    return (
      <dl className="BindingDetails container margin-normal">
        <dt key={secretName}>
          {secretName} <a onClick={this.openModal}>(show)</a>
        </dt>
        <TerminalModal
          modalIsOpen={this.state.modalIsOpen}
          closeModal={this.closeModal}
          title={`Secret: ${secretName}`}
          message={message}
        />
      </dl>
    );
  }

  public openModal = () => {
    this.setState({
      modalIsOpen: true,
    });
  };

  public closeModal = async () => {
    this.setState({
      modalIsOpen: false,
    });
  };
}

export default BindingDetails;
