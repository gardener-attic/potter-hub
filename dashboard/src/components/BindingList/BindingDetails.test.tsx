import { shallow } from "enzyme";
import * as React from "react";

import { IServiceBindingWithSecret } from "shared/ServiceBinding";
import TerminalModal from "../../components/TerminalModal";
import BindingDetails from "./BindingDetails";

const bindingWithSecret = {
  binding: {
    metadata: {
      name: "foo",
      namespace: "default",
      selfLink: "",
      uid: "",
      resourceVersion: "",
      creationTimestamp: "",
      finalizers: [],
      generation: 1,
    },
    spec: {
      externalID: "",
      instanceRef: {
        name: "foo",
      },
      secretName: "bar",
    },
    status: {
      asyncOpInProgress: false,
      currentOperation: "",
      reconciledGeneration: 1,
      operationStartTime: "",
      externalProperties: {},
      orphanMitigationInProgress: false,
      unbindStatus: "",
      conditions: [
        {
          lastTransitionTime: "1",
          type: "a type",
          status: "good",
          reason: "none",
          message: "everything okay here",
        },
      ],
    },
  },
  secret: {
    kind: "",
    apiVersion: "",
    metadata: {
      name: "bar",
      namespace: "default",
      selfLink: "",
      uid: "",
      resourceVersion: "",
      creationTimestamp: "",
      finalizers: [],
      generation: 1,
    },
    data: {
      mySecret: "Y29udGVudAo=",
    },
  },
} as IServiceBindingWithSecret;

it("renders information about the binding with a secret", () => {
  const wrapper = shallow(<BindingDetails {...bindingWithSecret} />);
  expect(wrapper.text()).toContain("bar (show)");
  expect(wrapper).toMatchSnapshot();
});

it("should show the secret content when clicking the button", () => {
  const wrapper = shallow(<BindingDetails {...bindingWithSecret} />);
  const button = wrapper.find("a");
  expect(button).toExist();

  expect((wrapper.state() as any).modalIsOpen).toBe(false);
  button.simulate("click");
  expect((wrapper.state() as any).modalIsOpen).toBe(true);
  expect(wrapper.find(TerminalModal).props().message).toContain("mySecret: content");
});
