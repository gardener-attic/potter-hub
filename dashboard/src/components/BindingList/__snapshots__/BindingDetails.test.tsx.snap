// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders information about the binding with a secret 1`] = `
<dl
  className="BindingDetails container margin-normal"
>
  <dt
    key="bar"
  >
    bar
     
    <a
      onClick={[Function]}
    >
      (show)
    </a>
  </dt>
  <TerminalModal
    closeModal={[Function]}
    message="mySecret: content

"
    modalIsOpen={false}
    title="Secret: bar"
  />
</dl>
`;
